source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -329,25 +337,34 @@ void DnsCacheImpl::finishResolve(const std::string& host,
   auto current_address = primary_host_info->host_info_->address();
   if (new_address != nullptr && (current_address == nullptr || *current_address != *new_address)) {
     if (!from_cache) {
-      addCacheEntry(host, new_address);
+      addCacheEntry(host, new_address, response.front().ttl_);
     }
-    // TODO(alyssawilk) don't immediately push cached entries to threads.
-    // Only serve stale entries if a configured resolve timeout has fired.
     ENVOY_LOG(debug, "host '{}' address has changed", host);
     primary_host_info->host_info_->setAddress(new_address);
     runAddUpdateCallbacks(host, primary_host_info->host_info_);
     address_changed = true;
     stats_.host_address_changed_.inc();
   }
 
-  if (first_resolve || address_changed) {
+  SystemTime now = main_thread_dispatcher_.timeSource().systemTime();
+  uint64_t ms_since_epoch =
+      std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()).count();
+  uint64_t stale_at = ms_since_epoch + std::chrono::duration_cast<std::chrono::seconds>(response.front().ttl_).count();
+  primary_host_info->host_info_->updateStale(stale_at);
```

Comment:
nit: It might make sense for this method to take the ttl directly and do the work in the method of coverting the ttl to whatever the internal representation is of "now + ttl"?